source/common/event/dispatcher_impl.cc

```diff
@@ -157,28 +154,25 @@ DispatcherImpl::createClientConnection(Network::Address::InstanceConstSharedPtr
                                                          std::move(transport_socket), options);
 }
 
+// Create DNS resolver based on the @param typed_dns_resolver_config, which could be
+// c-ares DNS resolver, Apple DNS resolver, or any other DNS resolver type.
 Network::DnsResolverSharedPtr DispatcherImpl::createDnsResolver(
-    const std::vector<Network::Address::InstanceConstSharedPtr>& resolvers,
-    const envoy::config::core::v3::DnsResolverOptions& dns_resolver_options) {
+    const envoy::config::core::v3::TypedExtensionConfig& typed_dns_resolver_config) {
+
   ASSERT(isThreadSafe());
-#ifdef __APPLE__
-  static bool use_apple_api_for_dns_lookups =
-      Runtime::runtimeFeatureEnabled("envoy.restart_features.use_apple_api_for_dns_lookups");
-  if (use_apple_api_for_dns_lookups) {
-    RELEASE_ASSERT(
-        resolvers.empty(),
-        "defining custom resolvers is not possible when using Apple APIs for DNS resolution. "
-        "Apple's API only allows overriding DNS resolvers via system settings. Delete resolvers "
-        "config or disable the envoy.restart_features.use_apple_api_for_dns_lookups runtime "
-        "feature.");
-    RELEASE_ASSERT(!dns_resolver_options.use_tcp_for_dns_lookups(),
-                   "using TCP for DNS lookups is not possible when using Apple APIs for DNS "
-                   "resolution. Apple' API only uses UDP for DNS resolution. Use UDP or disable "
-                   "the envoy.restart_features.use_apple_api_for_dns_lookups runtime feature.");
-    return std::make_shared<Network::AppleDnsResolverImpl>(*this, api_.rootScope());
-  }
-#endif
-  return std::make_shared<Network::DnsResolverImpl>(*this, resolvers, dns_resolver_options);
+
+  // typed_dns_resolver_config need to be populated.
+  ASSERT(!typed_dns_resolver_config.name().empty() &&
+         !typed_dns_resolver_config.typed_config().type_url().empty());
+
+  // Derive the DNS resolver factory from config.
+  // If the typed config is not registered or bogus, it will throw an exception.
+  // Since this function is only called in control plane, it is safe to throw an exception here.
```

Comment:
I'm not super comfortable with just documenting this; there should be an ASSERT indicating we are on the main thread, i.e. `ASSERT(Thread::MainThread::isMainThread());`.

Why can't a DNS resolver be created on a worker thread by a factory functino for a filter?

Reply:
Yes, a DNS resolver should be allowed to be created in worker thread by a factory function for a filter.
New changes are added to support this. More details see comments below.